File: src/ru/edu/vtisov/string/FindSmallestLetterGreaterThanTarget.java
package ru.edu.vtisov.string;

public class FindSmallestLetterGreaterThanTarget {
    public char nextGreatestLetter(char[] letters, char target) {
        int left = 0;
        int right = letters.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (target < letters[mid]) right = mid - 1;
            else left = mid + 1;
        }
        return letters[left % letters.length];
    }
}

File: src/ru/edu/vtisov/array/binarysearch/find-smallest-letter-greater-than-target.go
package binarysearch

func nextGreatestLetter(letters []byte, target byte) byte {
	i, j := 0, len(letters) // j за пределами массива
	for i < j {
		mid := i + (j-i)/2
		if letters[mid] <= target {
			i = mid + 1
		} else {
			j = mid
		}
	}
	return letters[i%len(letters)]
}

// func nextGreatestLetter(letters []byte, target byte) byte {
//     i, j := 0, len(letters)-1

//     for i < j {
//         mid := i+(j-i)/2
//         if letters[mid] <= target {
//             i = mid+1
//         } else {
//             j = mid
//         }
//     }
//     if letters[i] <= target {
//         return letters[0]
//     }
//     return letters[i]
// }
